cpp/src/components/scc_matrix.cuh

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <thrust/device_vector.h>
+#include <thrust/find.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/extrema.h>
+
+#include <type_traits>
+//
+
+//Convergence check logic;
+//
+/**
+ * @brief Provide convergence check logic for GEMM SCC via a device pointer
+ */
+struct CStableChecker
+{
+  explicit CStableChecker(int flag):
+    d_flag_(1, flag)
+  {
+  }
+
+  //hopefully might be cheaper than copying the value from device to host:
+  //
+  bool is_set(void) const
+  {
+    auto pos = thrust::find(d_flag_.begin(), d_flag_.end(), 1);
+    return (pos != d_flag_.end());
+  }
+
+  void set(int flag)
+  {
+    thrust::for_each(d_flag_.begin(), d_flag_.end(),
+                     [flag] __device__ (int& val){
+                       val = flag;
+                     });
+  }
+
+  int* get_ptr(void)
+  {
+    return d_flag_.data().get();
+  }
+private:
+  thrust::device_vector<int> d_flag_;
+};
+
+
+/**
+ * @brief SCC Algorithm
+ * (Adapted from John Gilbert's "Graph Algorithms in the Language of Linear Algebra")
+ *
+ * C = I | A | A^2 |...| A^k
+ * where matrix multiplication is through
+ * (combine, reduce) == (&, |) (bitwise ops)
+ * Then: X = C & transpose(C);
+ * apply get_labels(X);
+ */
+template<typename ByteT,
+         typename IndexT = int>
+struct SCC_Data
+{
+  SCC_Data(size_t nrows,
+           const IndexT* p_d_r_o, //row_offsets
+           const IndexT* p_d_c_i): //column indices
+    nrows_(nrows),
+    p_d_r_o_(p_d_r_o),
+    p_d_c_i_(p_d_c_i),
+    d_C(nrows*nrows, 0),
+    d_Cprev(nrows*nrows, 0)
+  {
+    init();
+  }
+
+  const thrust::device_vector<ByteT>& get_C(void) const
+  {
+    return d_C;
+  }
+
+  size_t nrows(void) const
+  {
+    return nrows_;
+  }
+
+  const IndexT* r_o(void) const
+  {
+    return p_d_r_o_;
+  }
+
+  const IndexT* c_i(void) const
+  {
+    return p_d_c_i_;
+  }
+  
+  //protected: cannot have device lambda inside protected memf
+  void init(void)
+  {    
+    //assert( d_Cprev.size() == nrows_ * nrows_ );
+    
+    //init d_Cprev to identity:
+    //
```

Comment:
Better delete the temporary commented out code.

Reply:
Ok, removed dead code.